File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0646_Maximum_Length_of_Pair_Chain.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Arrays;

public class Question_0646_Maximum_Length_of_Pair_Chain {
    public int findLongestChain(int[][] pairs) {
        Arrays.sort(pairs, (a, b) -> a[1] - b[1]);
        int cur = Integer.MIN_VALUE, res = 0;
        for (int[] pair: pairs) {
            if (cur < pair[0]) {
                cur = pair[1];
                res++;
            }
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0646_Maximum_Length_of_Pair_Chain.cc
#include <vector>

using namespace std;

class Solution {
 public:
  int findLongestChain(vector<vector<int>>& pairs) {
    sort(pairs.begin(), pairs.end(), cmp);
    int cur = pairs[0][0] - 1, res = 0;
    for (vector<int> p : pairs) {
      if (cur < p[0]) {
          cur = p[1];
          res++;
      }
    }
    return res;
  }

 private:
  static bool cmp(vector<int>& a, vector<int>&b) {
    return a[1] < b[1] || a[1] == b[1] && a[0] < b[0];
  }
};
